### ISSUE TITLE
Why is async function needed at all?

### ISSUE THREAD

**Issue by Rush** [tc39/proposal-async-await#88]
It seems to me that `async function` is just redundant. If the compiler sees `await` in function body, it could automatically wrap the code with appropriate generator+Promise.

After all, even without `async`/`await` we write many functions that simply return promises. Cannot we then assume that using `await` is a conscious enough choice  from the code writer perspective? From the user perspective, doing `import {someFunc} from 'foo';` does not say if the function is `async` or not, they need to read the docs or simply try it out. The thid perspective is the compiler .. and this is my guess my `async` qualifier exists but I couldn't find an official justification. It's probably easier to implement if the compiler can branch early to implement this different behavior but isn't less pain for a programmer a good justification to make it nicer? Our code bases will just end up with a soup of `async` in front of every `lambda` and `function`.

In short, let's make `async`/`await` an `await` proposal.

**Comment by RangerMauve** [comment]
I think that sometimes you'd get APIs that expect async functions, but you'd want your actual function body to not bother with any `await` calls.

**Comment by zenparsing** [comment]
Keep in mind that `await` is not a keyword outside of async functions.  So this is actually a valid function call:

``` js
function f() {
  await (something);
}
```

Among other reasons, we need the special form so that we can parse "await" as a keyword inside of it.

**Comment by Rush** [comment]
> I think that sometimes you'd get APIs that expect async functions, but you'd want your actual function body to not bother with any await calls.

Whatever async functions APIs expect, a normal Promise returning function would work too.

> Among other reasons, we need the special form so that we can parse "await" as a keyword inside of it.

Yes that's true but I'm pretty sure that it can be fixed so that `await` is a normal keyword. As I understand (correct me if I'm wrong) syntax-level compatibility is not be expected between major ES version releases?

**Comment by getify** [comment]
> syntax-level compatibility is not be expected between major ES version releases?

LOL. You should read [Backward Compatibility](https://en.wikipedia.org/wiki/Backward_compatibility). The web platform (incl/esp JS) is extremely loathe to break older code with new syntax updates.

**Comment by ljharb** [comment]
@Rush Don't forget that `global.Promise = somethingElse;` breaks your promise code, but won't break `async function` or `await`.

**Comment by sebmck** [comment]
Note that async functions are already stage 3. It would be very unreasonable and unproductive to suggest such drastic changes at this point.

**Comment by RangerMauve** [comment]
Also, I personally like seeing the `async` keyword right at the start of the definition because it's a cue that I'm dealing with async stuff. Whereas without it, I'd only clue into the fact that I'm dealing with an async function once I saw the `await` keyword, which might be hidden in between some spaghetti in certain cases.

**Comment by Rush** [comment]
> LOL. You should read Backward Compatibility. The web platform (incl/esp JS) is extremely loathe to break older code with new syntax updates.

No reason to ridicule and I would be gladly educated on why my suggested approach is impossible . I am not convinced that it necessarily has to break. 

> @Rush Don't forget that global.Promise = somethingElse; breaks your promise code, but won't break async function or await.

That's nice but if merely using `await` in an ordinary function simply upgraded that function to an `async` function this behavior could be retained.

> Also, I personally like seeing the async keyword right at the start of the definition because it's a cue that I'm dealing with async stuff. Whereas without it, I'd only clue into the fact that I'm dealing with an async function once I saw the await keyword, which might be hidden in between some spaghetti in certain cases.

I get ya however you will have this advantage only in your own code base and only in the same file. Any external interface, importing a library or some other file you will lose this async clue. Imagine writing:
`Promise.map(array, vallue => { foo1.......; foo2;` and then suddenly you realize that you need the `await`, you now need to rewind and add the `async`. I'd rather be able to use `await` immediately, just as I can simply return a `Promise` today. In my mind `await` is simply an implicit `Promise`.

Thanks guys for all your replies. I really appreciate it,

**Comment by getify** [comment]
> I am not convinced that it necessarily has to break.

Perhaps you missed @zenparsing's point.

``` js
function await() { console.log("woot"); }
function hello(x) { await(x); }
```

That program is entirely valid in current JS, and has a certain set of expected behavior. But that program will behave very differently simply by it running in a ES(2017ish) engine. That's a textbook definition of non-backwards-compatible breakage. Much to my chagrin, personally, there's a long established history of that kind of breakage being unacceptable.

To disambiguate that your program would want `await` to be treated as a keyword instead of as a function call, something has to tell JS that it's a different situation. That something has been decided to be the word "async" in front of the word "function".

Is that convincing now?

**Comment by bterlson** [comment]
Long ago we considered whether it was possible to `await` everywhere but we could not find a good way to do it without breaking existing usage of `await` as an identifier. I think a simpler example of a program that would break by turning `await` into a keyword is simply that - `await`, which must continue to parse as an identifier in non-async-function contexts. So it must be some sort of hybrid contextual keyword where it's an identifier sometimes and not others. Currently the line is very clear and has zero risk of backward compat issues. Moving to a subtler, more contextual keyword is rife with difficulty for both implementers and developers and is not something I'm interested in pursuing further (especially since this proposal is close to stage 4).

**Comment by Rush** [comment]
> Perhaps you missed @zenparsing's point.

Indeed I did, thank you! 

> Is that convincing now?

Yes that would break . Convinced. :) I was under an impression that await cannot be used with `await(x)` but only with `await x`. Obviously any expression can be wrapped with braces and still be an expression..

> Long ago we considered whether it was possible to await everywhere but we could not find a good way to do it without breaking existing usage of await as an identifier. I think a simpler example of a program that would break by turning await into a keyword is simply that - await, which must continue to parse as an identifier in non-async-function contexts. So it must be some sort of hybrid contextual keyword where it's an identifier sometimes and not others. Currently the line is very clear and has zero risk of backward compat issues. Moving to a subtler, more contextual keyword is rife with difficulty for both implementers and developers and is not something I'm interested in pursuing further (especially since this proposal is close to stage 4).

Thanks, it makes sense but let me dwell a little bit more though ...

Does the below mean it will be possible to do what I ask in the future? And if it is reserved to be a generic keyword one day, why not today? Some things got upgraded into keywords like `let` for example (looking at the list here https://mathiasbynens.be/notes/reserved-keywords if it is trustworthy).

http://ecma-international.org/ecma-262/6.0/

> 11.6.2.2 Future Reserved Words
> 
> The following tokens are reserved for used as keywords in future language extensions.
> 
> Syntax
> 
> FutureReservedWord ::
> enum
> await
> await is only treated as a FutureReservedWord when Module is the goal symbol of the syntactic grammar.
> 
> NOTE Use of the following tokens within strict mode code (see 10.2.1) is also reserved. That usage is restricted using static semantic restrictions (see 12.1.1) rather than the lexical grammar:
> 
> implements package protected  
> interface   private public

**Comment by bterlson** [comment]
> Does the below mean it will be possible to do what I ask in the future?

I'll never say never, but it doesn't seem likely.

> And if it is reserved to be a generic keyword one day, why not today?

There is no plan to make it a generic keyword due in part to the myriad problems raised in this thread.

> Some things got upgraded into keywords like  let  for example 

Let is easier because it's a statement. There were indeed breaking changes in upgrading let to a keyword but we were able to get convincing data beforehand that it was a safe change to make. Await is much harder.

**Comment by twome** [comment]
The rationale that treating `await` contextually *either* a variable or a keyword would be difficult for implementers makes sense to me. However, I think the this implicit `async` functionality could be incredibly useful for conciseness. Two questions/suggestions:

## Option 1: Future reserved word 

Could a future reserved word be added that would be semantically appropriate for this implicit `async` functionality proposed? 

### `wait` 

The asynchronous focus of JS means it's unlikely we'll ever need a synchronous `wait` command, and it would be *very* easy to create one using existing `await` and `setTimeout`. `wait` preserves an obvious link to `async/await`, and even implies that it's a shorthand, given that it's a substring of `await`. `async/await` both start with "a", which makes an easy mnemonic to distinguish them from `wait`.

### `abide`, `anticipate`
 
If you want to retain the first letter "a", these are both very descriptive.

### `stay`, `linger`, `bide`

These clearly communicate that the synchronous execution is going to pause at this point.

### `foretell`, `foresee`, `prepare`, `suppose`, `plan`

These imply that this is an event in the future, and one we can't be sure about the results of.

### `expect`, `assume`

Without the parser determining contextually whether these are variable names or keywords, these two would be annoying for automated tests which frequently use these names. 

---

## Option 2: ES module type 

If this implicit `async` rule only applied within ES modules (`type="module"`), could parsers correctly assume that `await` is already a keyword that would throw an exception if a user tried to assign it as a variable name?

**Comment by ljharb** [comment]
No new future reserved words can likely be added because they are already usable as variable names.

This repo is for a finished language feature; changing async/await at this time is not really practical.
